File: example/java/save-command/src/test/java/org/babyfish/jimmer/example/save/common/AbstractMutationWithTriggerTest.java
package org.babyfish.jimmer.example.save.common;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.event.TriggerType;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;

import java.util.ArrayList;
import java.util.List;

public class AbstractMutationWithTriggerTest extends AbstractMutationTest {

    private final List<String> events = new ArrayList<>();

    @Override
    protected void customize(JSqlClient.Builder builder) {
        builder.setTriggerType(TriggerType.TRANSACTION_ONLY);
    }

    @BeforeEach
    public void registerEventListeners() {
        sql().getTriggers().addEntityListener(e -> {
            events.add(
                    "The entity \"" +
                            e.getImmutableType() +
                            "\" is changed, " +
                            "old: " + e.getOldEntity() +
                            ", new: " + e.getNewEntity()
            );
        });
        sql().getTriggers().addAssociationListener(e -> {
            events.add(
                    "The association \"" +
                            e.getImmutableProp() +
                            "\" is changed, " +
                            "source id: " + e.getSourceId() +
                            ", detached target id: " + e.getDetachedTargetId() +
                            ", attached target id: " + e.getAttachedTargetId()
            );
        });
    }

    protected void assertEvents(String ... events) {
        int size = Math.min(this.events.size(), events.length);
        for (int i = 0; i < size; i++) {
            Assertions.assertEquals(
                    events[i],
                    this.events.get(i),
                    "events[i]: expected \"" +
                            events[i] +
                            "\", actual \"" +
                            this.events.get(i) +
                            "\""
            );
        }
        Assertions.assertEquals(
                events.length,
                this.events.size(),
                "Event count: expected " +
                        events.length +
                        ", actual " +
                        this.events.size()
        );
    }
}

File: example/kotlin/save-command-kt/src/test/kotlin/org/babyfish/jimmer/example/save/common/AbstractMutationWithTriggerTest.kt
package org.babyfish.jimmer.example.save.common

import org.babyfish.jimmer.sql.event.TriggerType
import org.babyfish.jimmer.sql.kt.KSqlClientDsl
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import kotlin.math.min

abstract class AbstractMutationWithTriggerTest : AbstractMutationTest() {

    private val events: MutableList<String> = ArrayList()

    override fun customize(dsl: KSqlClientDsl) {
        dsl.setTriggerType(TriggerType.TRANSACTION_ONLY)
    }

    @BeforeEach
    fun registerEventListeners() {

        events.clear()

        /*
         * If jimmer-spring-starter is used, it's unnecessary
         * to use triggers to register listeners explicitly,
         * because jimmer trigger events will be published
         * as spring events automatically
         */
        sql.triggers.addEntityListener { e ->
            events.add(
                "The entity \"" +
                    e.immutableType +
                    "\" is changed, " +
                    "old: " + e.oldEntity +
                    ", new: " + e.newEntity
            )
        }
        sql.triggers.addAssociationListener { e ->
            events.add(
                "The association \"" +
                    e.immutableProp.toString() +
                    "\" is changed, " +
                    "source id: " + e.sourceId +
                    ", detached target id: " + e.detachedTargetId +
                    ", attached target id: " + e.attachedTargetId
            )
        }
    }

    protected fun assertEvents(vararg events: String) {
        val size = min(this.events.size, events.size)
        for (i in 0 until size) {
            Assertions.assertEquals(
                events[i],
                this.events[i],
                "events[i]: expected \"" +
                    events[i] +
                    "\", actual \"" +
                    this.events[i] +
                    "\""
            )
        }
        Assertions.assertEquals(
            events.size,
            this.events.size,
            ("Event count: expected " +
                events.size +
                ", actual " +
                this.events.size)
        )
    }
}